Virtio over shmem

YuanyuanZhao609 opened the pull request:
Add virtio over ivshmem to acrn and QNX platform:
1. Screencast on acrn platform work successfully.
2. Virtio gpu and Virtio net over ivshmem on QNX platform work successfully.
3. Virtio blk work normally.

Tracked-On: OAM-119803

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

xyzhao2018 commented on arch/x86/kernel/cpu/common.c:
```diff
@@ -990,20 +991,23 @@ static void init_speculation_control(struct cpuinfo_x86 *c)
 		set_cpu_cap(c, X86_FEATURE_STIBP);
-		set_cpu_cap(c, X86_FEATURE_MSR_SPEC_CTRL);
+		if (!hypervisor_is_type(X86_HYPER_QNX))
+			set_cpu_cap(c, X86_FEATURE_MSR_SPEC_CTRL);
```
Why we also remove it for AMD processor?

YuanyuanZhao609 replied:
Yes.

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

JeevakaPrabu reviewed (changes requested):
jbadrapp@intel-S2600STB:~/celadon_a14/kernel/linux-intel-lts2022$ ./scripts/checkpatch.pl 0001-x86-qnx-Add-support-for-Linux-guests-on-QNX-hypervis.patch
Traceback (most recent call last):
  File "scripts/spdxcheck.py", line 11, in <module>
    import git
ModuleNotFoundError: No module named 'git'
WARNING: please write a help paragraph that fully describes the config symbol
#30: FILE: arch/x86/Kconfig:901:
+config QNX_GUEST
+       bool "QNX Guest support"
+       depends on X86_64
+       help
+         This option allows to run Linux as guest on QNX hypervisor.
+

WARNING: added, moved or deleted file(s), does MAINTAINERS need updating?
#87:
new file mode 100644

WARNING: Missing a blank line after declarations
#108: FILE: arch/x86/kernel/cpu/qnx.c:17:
+               uint32_t eax, signature[2], edx;
+               cpuid(QNX_CPUID_FEATURES, &eax, &signature[0], &signature[1], &edx);

total: 0 errors, 3 warnings, 88 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

0001-x86-qnx-Add-support-for-Linux-guests-on-QNX-hypervis.patch has style problems, please review.

NOTE: If any of the errors are false positives, please report
      them to the maintainer, see CHECKPATCH in MAINTAINERS.






jbadrapp@intel-S2600STB:~/celadon_a14/kernel/linux-intel-lts2022$ ./scripts/checkpatch.pl 0002-virtio_shmem-virtio-over-shared-memory-front-end-dev.patch
Traceback (most recent call last):
  File "scripts/spdxcheck.py", line 11, in <module>
    import git
ModuleNotFoundError: No module named 'git'
Traceback (most recent call last):
  File "scripts/spdxcheck.py", line 11, in <module>
    import git
ModuleNotFoundError: No module named 'git'
Traceback (most recent call last):
  File "scripts/spdxcheck.py", line 11, in <module>
    import git
ModuleNotFoundError: No module named 'git'
Traceback (most recent call last):
  File "scripts/spdxcheck.py", line 11, in <module>
    import git
ModuleNotFoundError: No module named 'git'
Traceback (most recent call last):
  File "scripts/spdxcheck.py", line 11, in <module>
    import git
ModuleNotFoundError: No module named 'git'
Traceback (most recent call last):
  File "scripts/spdxcheck.py", line 11, in <module>
    import git
ModuleNotFoundError: No module named 'git'
WARNING: please write a help paragraph that fully describes the config symbol
#44: FILE: drivers/virtio/Kconfig:176:
+menuconfig VIRTIO_IVSHMEM
+       tristate "Driver for ivshmem-based virtio front-end devices"
+       depends on PCI && !HIGHMEM
+       select VIRTIO
+       select DMA_OPS
+       select ZONE_DEVICE
+       help
+         This provides virtio front-end devices via ivshmem shared memory
+         devices.
+
+         If unsure, say 'N'.
+

WARNING: please write a help paragraph that fully describes the config symbol
#58: FILE: drivers/virtio/Kconfig:190:
+config VIRTIO_IVSHMEM_IVSHMEM
+       tristate "Driver for QEMU IVSHMEM"
+       depends on VIRTIO_IVSHMEM
+       help
+         Driver for QEMU IVSHMEM.
+

WARNING: please write a help paragraph that fully describes the config symbol
#64: FILE: drivers/virtio/Kconfig:196:
+config VIRTIO_IVSHMEM_GUEST_SHM
+       tristate "Driver for QNX guest shm"
+       depends on VIRTIO_IVSHMEM
+       help
+         Driver for QNX guest shm.
+

WARNING: please write a help paragraph that fully describes the config symbol
#70: FILE: drivers/virtio/Kconfig:202:
+config VIRTIO_IVSHMEM_DEBUG
+       tristate "Driver for QEMU IVSHMEM"
+       depends on VIRTIO_IVSHMEM
+       help
+        This driver is used to debug virtio-ivshmem performance statistics.
+
+        If unsure, say N.
+

WARNING: added, moved or deleted file(s), does MAINTAINERS need updating?
#96:
new file mode 100644

WARNING: Prefer 'unsigned int' to bare use of 'unsigned'
#127: FILE: drivers/virtio/virtio_guest_shm.c:27:
+       unsigned attach_list;

WARNING: Prefer 'unsigned int' to bare use of 'unsigned'
#142: FILE: drivers/virtio/virtio_guest_shm.c:42:
+       unsigned i;

WARNING: braces {} are not necessary for single statement blocks
#148: FILE: drivers/virtio/virtio_guest_shm.c:48:
+               if (attach_list_changes & (1u << i)) {
+                       dev_info(&vi_dev->pci_dev->dev, "client %2u %stached", i, (attach_list & (1u << i)) ? "at" : "de");
+               }

WARNING: line length of 123 exceeds 100 columns
#149: FILE: drivers/virtio/virtio_guest_shm.c:49:
+                       dev_info(&vi_dev->pci_dev->dev, "client %2u %stached", i, (attach_list & (1u << i)) ? "at" : "de");

WARNING: line length of 112 exceeds 100 columns
#195: FILE: drivers/virtio/virtio_guest_shm.c:95:
+                       (unsigned long long)GUEST_SHM_SIGNATURE, (unsigned long long) vi_data->fact->signature);

WARNING: line length of 159 exceeds 100 columns
#211: FILE: drivers/virtio/virtio_guest_shm.c:111:
+       dev_info(&pci_dev->dev, "virtio%ld creation size %x, paddr=%llx, irq: %d", idx, vi_data->fact->size, (unsigned long long)paddr, vi_data->fact->vector);

WARNING: line length of 140 exceeds 100 columns
#224: FILE: drivers/virtio/virtio_guest_shm.c:124:
+       section_sz = vi_data->fact->size * 0x1000 - ((section_addr - (paddr + (idx << 32)) - 0x1000 + (min_align - 1))  & ~(min_align - 1));

WARNING: braces {} are not necessary for single statement blocks
#235: FILE: drivers/virtio/virtio_guest_shm.c:135:
+       if (ret) {
+               goto err_enable;
+       }

WARNING: line length of 101 exceeds 100 columns
#248: FILE: drivers/virtio/virtio_guest_shm.c:148:
+       vi_dev->shmem_sz_used = (vi_dev->virtio_header->size + (chunk_size - 1)) & ~(chunk_size - 1);

WARNING: Prefer "GPL" over "GPL v2" - see commit bf7fbeeae6db ("module: Cure the MODULE_LICENSE "GPL" vs. "GPL v2" bogosity")
#295: FILE: drivers/virtio/virtio_guest_shm.c:195:
+MODULE_LICENSE("GPL v2");

WARNING: braces {} are not necessary for single statement blocks
#424: FILE: drivers/virtio/virtio_ivshmem.c:123:
+       if (ret) {
+               goto err_enable;
+       }

WARNING: line length of 101 exceeds 100 columns
#436: FILE: drivers/virtio/virtio_ivshmem.c:135:
+       vi_dev->shmem_sz_used = (vi_dev->virtio_header->size + (chunk_size - 1)) & ~(chunk_size - 1);

WARNING: Prefer "GPL" over "GPL v2" - see commit bf7fbeeae6db ("module: Cure the MODULE_LICENSE "GPL" vs. "GPL v2" bogosity")
#482: FILE: drivers/virtio/virtio_ivshmem.c:181:
+MODULE_LICENSE("GPL v2");

WARNING: braces {} are not necessary for single statement blocks
#555: FILE: drivers/virtio/virtio_shmem.c:67:
+       while (READ_ONCE(vi_dev->virtio_header->write_transaction)) {
+               cpu_relax();
+       }

WARNING: Do not crash the kernel unless it is absolutely unavoidable--use WARN_ON_ONCE() plus recovery code (if feasible) instead of BUG() or variants
#578: FILE: drivers/virtio/virtio_shmem.c:90:
+               BUG();

WARNING: memory barrier without comment
#579: FILE: drivers/virtio/virtio_shmem.c:91:
+       virt_wmb();

WARNING: memory barrier without comment
#582: FILE: drivers/virtio/virtio_shmem.c:94:
+       virt_wmb();

WARNING: Prefer 'unsigned int' to bare use of 'unsigned'
#618: FILE: drivers/virtio/virtio_shmem.c:130:
+static void vi_get(struct virtio_device *vdev, unsigned offset,

WARNING: Prefer 'unsigned int' to bare use of 'unsigned'
#619: FILE: drivers/virtio/virtio_shmem.c:131:
+                  void *buf, unsigned len)

WARNING: Do not crash the kernel unless it is absolutely unavoidable--use WARN_ON_ONCE() plus recovery code (if feasible) instead of BUG() or variants
#643: FILE: drivers/virtio/virtio_shmem.c:155:
+               BUG();

WARNING: Prefer 'unsigned int' to bare use of 'unsigned'
#647: FILE: drivers/virtio/virtio_shmem.c:159:
+static void vi_set(struct virtio_device *vdev, unsigned offset,

WARNING: Prefer 'unsigned int' to bare use of 'unsigned'
#648: FILE: drivers/virtio/virtio_shmem.c:160:
+                  const void *buf, unsigned len)

WARNING: Do not crash the kernel unless it is absolutely unavoidable--use WARN_ON_ONCE() plus recovery code (if feasible) instead of BUG() or variants
#666: FILE: drivers/virtio/virtio_shmem.c:178:
+               BUG();

WARNING: please, no space before tabs
#680: FILE: drivers/virtio/virtio_shmem.c:192:
+^I^I/* ^Idev_err_ratelimited(&vi_dev->pci_dev->dev, */$

WARNING: please, no space before tabs
#681: FILE: drivers/virtio/virtio_shmem.c:193:
+^I^I/* ^I^I^I    "backend failed!"); */$

WARNING: please, no space before tabs
#682: FILE: drivers/virtio/virtio_shmem.c:194:
+^I^I/* ^Ireturn 0; */$

WARNING: Do not crash the kernel unless it is absolutely unavoidable--use WARN_ON_ONCE() plus recovery code (if feasible) instead of BUG() or variants
#703: FILE: drivers/virtio/virtio_shmem.c:215:
+       BUG_ON(status == 0);

WARNING: memory barrier without comment
#768: FILE: drivers/virtio/virtio_shmem.c:280:
+       virt_wmb();

WARNING: memory barrier without comment
#780: FILE: drivers/virtio/virtio_shmem.c:292:
+               virt_wmb();

WARNING: memory barrier without comment
#796: FILE: drivers/virtio/virtio_shmem.c:308:
+               virt_wmb();

WARNING: Prefer 'unsigned int' to bare use of 'unsigned'
#956: FILE: drivers/virtio/virtio_shmem.c:468:
+static int vi_find_vqs(struct virtio_device *vdev, unsigned nvqs,

WARNING: Prefer printk_ratelimited or pr_<level>_ratelimited to printk_ratelimit
#1130: FILE: drivers/virtio/virtio_shmem.c:642:
+               if (!(attrs & DMA_ATTR_NO_WARN) && printk_ratelimit())

ERROR: code indent should use tabs where possible
#1205: FILE: drivers/virtio/virtio_shmem.c:717:
+            ((dir == DMA_FROM_DEVICE) || (dir == DMA_BIDIRECTIONAL)))$

WARNING: please, no spaces at the start of a line
#1205: FILE: drivers/virtio/virtio_shmem.c:717:
+            ((dir == DMA_FROM_DEVICE) || (dir == DMA_BIDIRECTIONAL)))$

WARNING: line length of 114 exceeds 100 columns
#1314: FILE: drivers/virtio/virtio_shmem.c:826:
+                       vi_dev->dma_alloc_cnt - vi_dev->dma_map_cnt, vi_dev->dma_map_cnt, vi_dev->dma_map_sg_cnt);

WARNING: Prefer "GPL" over "GPL v2" - see commit bf7fbeeae6db ("module: Cure the MODULE_LICENSE "GPL" vs. "GPL v2" bogosity")
#1409: FILE: drivers/virtio/virtio_shmem.c:921:
+MODULE_LICENSE("GPL v2");

ERROR: code indent should use tabs where possible
#1431: FILE: drivers/virtio/virtio_shmem.h:16:
+        union {$

WARNING: please, no spaces at the start of a line
#1431: FILE: drivers/virtio/virtio_shmem.h:16:
+        union {$

WARNING: Missing or malformed SPDX-License-Identifier tag in line 1
#1509: FILE: include/linux/guest_shm.h:1:
+/*

WARNING: line length of 128 exceeds 100 columns
#1586: FILE: include/linux/guest_shm.h:78:
+       GSS_CLIENT_MAX,                 /**< creation failed due to region already being used by the maximum number of guests */

WARNING: line length of 135 exceeds 100 columns
#1616: FILE: include/linux/guest_shm.h:108:
+       uint32_t                status; /**< lower 16 bits: pending notification bitset, upper 16 bits: current active clients (R/O) */

WARNING: Use of volatile is usually wrong: see Documentation/process/volatile-considered-harmful.rst
#1624: FILE: include/linux/guest_shm.h:116:
+guest_shm_create(volatile struct guest_shm_factory *const __factory, unsigned const __size) {

WARNING: Block comments use a trailing */ on a separate line
#1627: FILE: include/linux/guest_shm.h:119:
+        * necessary bits (e.g. setting the name of the region) */

ERROR: space prohibited after that open parenthesis '('
#1628: FILE: include/linux/guest_shm.h:120:
+       asm volatile( "" ::: "memory");

ERROR: space prohibited after that open parenthesis '('
#1630: FILE: include/linux/guest_shm.h:122:
+       asm volatile( "" ::: "memory");

WARNING: Use of volatile is usually wrong: see Documentation/process/volatile-considered-harmful.rst
#1635: FILE: include/linux/guest_shm.h:127:
+guest_shm_find(volatile struct guest_shm_factory *const __factory, unsigned const __find_num) {

WARNING: Block comments use a trailing */ on a separate line
#1638: FILE: include/linux/guest_shm.h:130:
+        * necessary bits (e.g. setting the name of the region) */

ERROR: space prohibited after that open parenthesis '('
#1639: FILE: include/linux/guest_shm.h:131:
+       asm volatile( "" ::: "memory");

ERROR: space prohibited after that open parenthesis '('
#1641: FILE: include/linux/guest_shm.h:133:
+       asm volatile( "" ::: "memory");

WARNING: please, no spaces at the start of a line
#1708: FILE: include/linux/virtio_shm.h:16:
+    return ERR_PTR(-EINVAL);$

WARNING: please, no spaces at the start of a line
#1712: FILE: include/linux/virtio_shm.h:20:
+    return;$

WARNING: adding a line without newline at end of file
#1716: FILE: include/linux/virtio_shm.h:24:
+#endif /* _DRIVERS_VIRTIO_SHMEM_H */

total: 6 errors, 51 warnings, 1629 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

NOTE: Whitespace errors detected.
      You may wish to use scripts/cleanpatch or scripts/cleanfile

0002-virtio_shmem-virtio-over-shared-memory-front-end-dev.patch has style problems, please review.

NOTE: If any of the errors are false positives, please report
      them to the maintainer, see CHECKPATCH in MAINTAINERS.


jbadrapp@intel-S2600STB:~/celadon_a14/kernel/linux-intel-lts2022$ ./scripts/checkpatch.pl 0003-ivshm-ivsh-memory-region-drivers.patch
Traceback (most recent call last):
  File "scripts/spdxcheck.py", line 11, in <module>
    import git
ModuleNotFoundError: No module named 'git'
Traceback (most recent call last):
  File "scripts/spdxcheck.py", line 11, in <module>
    import git
ModuleNotFoundError: No module named 'git'
Traceback (most recent call last):
  File "scripts/spdxcheck.py", line 11, in <module>
    import git
ModuleNotFoundError: No module named 'git'
Traceback (most recent call last):
  File "scripts/spdxcheck.py", line 11, in <module>
    import git
ModuleNotFoundError: No module named 'git'
Traceback (most recent call last):
  File "scripts/spdxcheck.py", line 11, in <module>
    import git
ModuleNotFoundError: No module named 'git'
Traceback (most recent call last):
  File "scripts/spdxcheck.py", line 11, in <module>
    import git
ModuleNotFoundError: No module named 'git'
Traceback (most recent call last):
  File "scripts/spdxcheck.py", line 11, in <module>
    import git
ModuleNotFoundError: No module named 'git'
WARNING: added, moved or deleted file(s), does MAINTAINERS need updating?
#56:
new file mode 100644

WARNING: line length of 112 exceeds 100 columns
#163: FILE: drivers/ivshm/ivshm.c:59:
+static int ivshm_major_init(const char *name, const struct file_operations *fops, int *major, struct cdev **dev)

WARNING: line length of 111 exceeds 100 columns
#437: FILE: drivers/ivshm/ivshm.c:333:
+               err = remap_pfn_range(vma, vma->vm_start, pfn, vma->vm_end - vma->vm_start, vma->vm_page_prot);

WARNING: line length of 108 exceeds 100 columns
#439: FILE: drivers/ivshm/ivshm.c:335:
+               WARN_ONCE(1, "Shared memory region at 0x%llx won't integrate with udmabuf.", iregion->base);

WARNING: line length of 117 exceeds 100 columns
#446: FILE: drivers/ivshm/ivshm.c:342:
+static int ivshm_region_add_listener(struct ivshm_region_user *user, struct ivshm_region *iregion, unsigned long arg)

WARNING: line length of 115 exceeds 100 columns
#493: FILE: drivers/ivshm/ivshm.c:389:
+               return put_user(iregion->idev->info->dev_mmio_len, (unsigned long long __user *)arg) ? -EFAULT : 0;

WARNING: 'registeration' may be misspelled - perhaps 'registration'?
#559: FILE: drivers/ivshm/ivshm.c:455:
+ * Device registeration
           ^^^^^^^^^^^^^

WARNING: void function return statements are not generally useful
#676: FILE: drivers/ivshm/ivshm.c:572:
+       return;
+}

WARNING: Prefer strscpy, strscpy_pad, or __nonstring over strncpy - see: https://github.com/KSPP/linux/issues/90
#692: FILE: drivers/ivshm/ivshm.c:588:
+       strncpy(iregion->name, name, IVSHM_REGION_NAME_LEN);

WARNING: line length of 108 exceeds 100 columns
#784: FILE: drivers/ivshm/ivshm.c:680:
+       ret = ivshm_major_init("ivshm_region", &ivshm_region_fops, &ivshm_region_major, &ivshm_region_cdev);

WARNING: Prefer [subsystem eg: netdev]_err([subsystem]dev, ... then dev_err(dev, ... then pr_err(...  to printk(KERN_ERR ...
#790: FILE: drivers/ivshm/ivshm.c:686:
+               printk(KERN_ERR "class_register failed for ivshm\n");

WARNING: Prefer [subsystem eg: netdev]_err([subsystem]dev, ... then dev_err(dev, ... then pr_err(...  to printk(KERN_ERR ...
#796: FILE: drivers/ivshm/ivshm.c:692:
+               printk(KERN_ERR "class_register failed for ivshm_region\n");

WARNING: Prefer "GPL" over "GPL v2" - see commit bf7fbeeae6db ("module: Cure the MODULE_LICENSE "GPL" vs. "GPL v2" bogosity")
#837: FILE: drivers/ivshm/ivshm.c:733:
+MODULE_LICENSE("GPL v2");

WARNING: line length of 109 exceeds 100 columns
#1025: FILE: drivers/ivshm/ivshm_guest_shm.c:78:
+                       (unsigned long long)GUEST_SHM_SIGNATURE, (unsigned long long) idev->fact->signature);

WARNING: line length of 121 exceeds 100 columns
#1039: FILE: drivers/ivshm/ivshm_guest_shm.c:92:
+       dev_info(&pdev->dev, "virtio%ld creation size %x, paddr=%llx", idx, idev->fact->size, (unsigned long long)paddr);

WARNING: line length of 156 exceeds 100 columns
#1060: FILE: drivers/ivshm/ivshm_guest_shm.c:113:
+       idev->info.dev_mmio_len = idev->fact->size * 0x1000 - ((idev->info.dev_mmio - (paddr + (idx << 32)) - 0x1000 + (min_align - 1)) & ~(min_align - 1));

WARNING: line length of 138 exceeds 100 columns
#1061: FILE: drivers/ivshm/ivshm_guest_shm.c:114:
+       dev_info(&pdev->dev, "min_align: 0x%lx, mmio: 0x%llx, mmio_len: 0x%llx", min_align, idev->info.dev_mmio, idev->info.dev_mmio_len);

WARNING: line length of 143 exceeds 100 columns
#1063: FILE: drivers/ivshm/ivshm_guest_shm.c:116:
+       err = ivshm_register_region(idev->info.ivshm_dev, "default", idev->info.dev_mmio, idev->info.dev_mmio_len, 1 /* use INTx */, &iregion);

WARNING: Prefer "GPL" over "GPL v2" - see commit bf7fbeeae6db ("module: Cure the MODULE_LICENSE "GPL" vs. "GPL v2" bogosity")
#1109: FILE: drivers/ivshm/ivshm_guest_shm.c:162:
+MODULE_LICENSE("GPL v2");

WARNING: line length of 105 exceeds 100 columns
#1181: FILE: drivers/ivshm/ivshm_ivshmem.c:64:
+               err = request_irq(idev->msix_entries[i].vector, ivshmem_irq_handler, 0, "ivshmem", idev);

WARNING: Missing a blank line after declarations
#1184: FILE: drivers/ivshm/ivshm_ivshmem.c:67:
+                       int j;
+                       for (j = 0; j < i; j++) {

WARNING: braces {} are not necessary for single statement blocks
#1184: FILE: drivers/ivshm/ivshm_ivshmem.c:67:
+                       for (j = 0; j < i; j++) {
+                               free_irq(idev->msix_entries[i].vector, idev);
+                       }

WARNING: line length of 117 exceeds 100 columns
#1235: FILE: drivers/ivshm/ivshm_ivshmem.c:118:
+       err = ivshm_register_region(idev->info.ivshm_dev, "default", start, len, pci_msix_vec_count(pdev), &iregion);

WARNING: Prefer "GPL" over "GPL v2" - see commit bf7fbeeae6db ("module: Cure the MODULE_LICENSE "GPL" vs. "GPL v2" bogosity")
#1285: FILE: drivers/ivshm/ivshm_ivshmem.c:168:
+MODULE_LICENSE("GPL v2");

total: 0 errors, 24 warnings, 1239 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

0003-ivshm-ivsh-memory-region-drivers.patch has style problems, please review.

NOTE: If any of the errors are false positives, please report
      them to the maintainer, see CHECKPATCH in MAINTAINERS.


jbadrapp@intel-S2600STB:~/celadon_a14/kernel/linux-intel-lts2022$ ./scripts/checkpatch.pl 0006-QNX-guest_shm-hardcode-physical-address-of-shared-me.patch
WARNING: line length of 166 exceeds 100 columns
#40: FILE: drivers/ivshm/ivshm_guest_shm.c:113:
+       idev->info.dev_mmio_len = idev->fact->size * 0x1000 - ((idev->info.dev_mmio - (GUEST_SHM_PADDR + (idx << 32)) - 0x1000 + (min_align - 1)) & ~(min_align - 1));

WARNING: line length of 150 exceeds 100 columns
#56: FILE: drivers/virtio/virtio_guest_shm.c:124:
+       section_sz = vi_data->fact->size * 0x1000 - ((section_addr - (GUEST_SHM_PADDR + (idx << 32)) - 0x1000 + (min_align - 1))  & ~(min_align - 1));

total: 0 errors, 2 warnings, 29 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

0006-QNX-guest_shm-hardcode-physical-address-of-shared-me.patch has style problems, please review.

NOTE: If any of the errors are false positives, please report
      them to the maintainer, see CHECKPATCH in MAINTAINERS.


jbadrapp@intel-S2600STB:~/celadon_a14/kernel/linux-intel-lts2022$ ./scripts/checkpatch.pl 0007-QNX-WA-APIC-Timer-clock-accuracy.patch
WARNING: Prefer a maximum 75 chars per line (possible unwrapped commit description?)
#6:
WA QNX APIC LVT Timer clock accuracy issue by scaling APIC Timer Initial Count by setting

total: 0 errors, 1 warnings, 33 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

0007-QNX-WA-APIC-Timer-clock-accuracy.patch has style problems, please review.

NOTE: If any of the errors are false positives, please report
      them to the maintainer, see CHECKPATCH in MAINTAINERS.


jbadrapp@intel-S2600STB:~/celadon_a14/kernel/linux-intel-lts2022$ ./scripts/checkpatch.pl 0008-QNX-pci-MMIO-window-for-VF-passthrough.patch
ERROR: do not initialise statics to 0
#36: FILE: drivers/acpi/pci_root.c:899:
+static phys_addr_t pci_root_upper_base = 0;

ERROR: do not initialise statics to 0
#37: FILE: drivers/acpi/pci_root.c:900:
+static phys_addr_t pci_root_upper_limit = 0;

WARNING: simple_strtoull is obsolete, use kstrtoull instead
#44: FILE: drivers/acpi/pci_root.c:907:
+       value = simple_strtoull(str, &cur, 0);

WARNING: line length of 114 exceeds 100 columns
#48: FILE: drivers/acpi/pci_root.c:911:
+               printk(KERN_INFO "PCI PATCH CFG: Base of upper PCI MMIO window -> 0x%llx\n", pci_root_upper_base);

WARNING: Prefer [subsystem eg: netdev]_info([subsystem]dev, ... then dev_info(dev, ... then pr_info(...  to printk(KERN_INFO ...
#48: FILE: drivers/acpi/pci_root.c:911:
+               printk(KERN_INFO "PCI PATCH CFG: Base of upper PCI MMIO window -> 0x%llx\n", pci_root_upper_base);

WARNING: simple_strtoull is obsolete, use kstrtoull instead
#61: FILE: drivers/acpi/pci_root.c:924:
+       value = simple_strtoull(str, &cur, 0);

WARNING: line length of 116 exceeds 100 columns
#65: FILE: drivers/acpi/pci_root.c:928:
+               printk(KERN_INFO "PCI PATCH CFG: Limit of upper PCI MMIO window -> 0x%llx\n", pci_root_upper_limit);

WARNING: Prefer [subsystem eg: netdev]_info([subsystem]dev, ... then dev_info(dev, ... then pr_info(...  to printk(KERN_INFO ...
#65: FILE: drivers/acpi/pci_root.c:928:
+               printk(KERN_INFO "PCI PATCH CFG: Limit of upper PCI MMIO window -> 0x%llx\n", pci_root_upper_limit);

WARNING: line length of 114 exceeds 100 columns
#102: FILE: drivers/acpi/pci_root.c:972:
+                               (entry->res->flags & IORESOURCE_MEM) && entry->res->start > pci_root_upper_base) {

total: 2 errors, 7 warnings, 99 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

0008-QNX-pci-MMIO-window-for-VF-passthrough.patch has style problems, please review.

NOTE: If any of the errors are false positives, please report
      them to the maintainer, see CHECKPATCH in MAINTAINERS.


jbadrapp@intel-S2600STB:~/celadon_a14/kernel/linux-intel-lts2022$ ./scripts/checkpatch.pl 0009-QNX-pci-PCI-config-for-VF-passthrough.patch
ERROR: do not initialise statics to 0
#43: FILE: drivers/pci/access.c:46:
+static int config_override_rule_num = 0;

WARNING: Prefer [subsystem eg: netdev]_info([subsystem]dev, ... then dev_info(dev, ... then pr_info(...  to printk(KERN_INFO ...
#71: FILE: drivers/pci/access.c:74:
+       printk(KERN_INFO "PCI PATCH CFG: PCI bus 00:%02x.%d config register 0x%x -> 0x%x\n",

WARNING: line length of 146 exceeds 100 columns
#92: FILE: drivers/pci/access.c:95:
+                       dev_info(&bus->dev, "PCI PATCH: Config register %02x.%d : 0x%x -> 0x%x\n", PCI_SLOT(devfn), PCI_FUNC(devfn), where, *val);

WARNING: void function return statements are not generally useful
#101: FILE: drivers/pci/access.c:104:
+       return;
+}

ERROR: do not initialise statics to 0
#142: FILE: drivers/pci/probe.c:178:
+static int bar_override_rule_num = 0;

WARNING: simple_strtoull is obsolete, use kstrtoull instead
#164: FILE: drivers/pci/probe.c:200:
+               addr = simple_strtoull(str, &str, 0);

WARNING: simple_strtoull is obsolete, use kstrtoull instead
#167: FILE: drivers/pci/probe.c:203:
+               size = simple_strtoull(str, &str, 0);

WARNING: Prefer [subsystem eg: netdev]_info([subsystem]dev, ... then dev_info(dev, ... then pr_info(...  to printk(KERN_INFO ...
#187: FILE: drivers/pci/probe.c:223:
+       printk(KERN_INFO "PCI PATCH CFG: PCI bus 00:%02x.%d bar %d (%d bit) -> base 0x%llx, size 0x%llx\n",

WARNING: line length of 136 exceeds 100 columns
#218: FILE: drivers/pci/probe.c:254:
+                               pci_info(dev, "PCI PATCH: BAR %d: val = 0x%x, sz = 0x%x\n", (pos - PCI_BASE_ADDRESS_0) >> 2, *val, *sz);

total: 2 errors, 7 warnings, 236 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

0009-QNX-pci-PCI-config-for-VF-passthrough.patch has style problems, please review.

NOTE: If any of the errors are false positives, please report
      them to the maintainer, see CHECKPATCH in MAINTAINERS.


jbadrapp@intel-S2600STB:~/celadon_a14/kernel/linux-intel-lts2022$ ./scripts/checkpatch.pl 0011-QNX-i915-provisioning-VFs.patch
WARNING: Prefer a maximum 75 chars per line (possible unwrapped commit description?)
#6:
Provisioning VFs right after the driver is probed has a probability (~80%) to

WARNING: Do not crash the kernel unless it is absolutely unavoidable--use WARN_ON_ONCE() plus recovery code (if feasible) instead of BUG() or variants
#144: FILE: drivers/gpu/drm/i915/i915_sriov.c:128:
+               GEM_BUG_ON(!dev_is_pf(dev));

total: 0 errors, 2 warnings, 197 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

0011-QNX-i915-provisioning-VFs.patch has style problems, please review.

NOTE: If any of the errors are false positives, please report
      them to the maintainer, see CHECKPATCH in MAINTAINERS.

As per commit message, virtio_ivshmem.c is for ACRN hypervisor based VM whereas in Kconfig Driver for QEMU IVSHMEM

In drivers/virtio/virtio_guest_shm.c and drivers/virtio/virtio_ivshmem.c
    - In probe function, vi_dev is not freed on pcim_enable_device failure.
    - In probe function, vi_data is not freed neither in err_enable nor in err label
In virtio_ivshmem_remove function, vi_dev>priv not freed?

Have we modified the file drivers/virtio/virtio_guest_shm.c? If so whats the change we did in addition to the one done by Jan Kiszka <jan.kiszka@siemens.com>??

In drivers/ivshm/ivshm.c, 
- cdev is not freed on failures in ivshm_major_init 

In drivers/ivshm/ivshm_guest_shm.c,
- In probe function, idev is not freed on all failure exits
- In ivshmem_init_msix function, idev->msix_entries is not freed on all failure exits
- Check whether the registered regions are also released on all possible failure exits


May i know the reason behind the change,
param(unsigned int, max_vfs, ~0, 0400) \

Comment on commit message:
- Summary should be <72 characters and should be only 1 line
- Limit the max number of characters in the commit message to 72characters/line
- For workaround patches, please prefix [W/A]
- For patches which needs to be reverted in future after proper fix, commit message needs to be prefixed with [REVERT]
- Commit summary not proper for 13th patch 
Subject: [PATCH 13/14] =?UTF-8?q?QNX:=20x86=EF=BC=9Aremoving=20unsupported?=
 =?UTF-8?q?=20feature?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

YuanyuanZhao609 commented:
hi Jeevaka, thank you for your comments.

All patches checked with ./scripts/checkpatch.pl, and errors have resolved. 

    'As per commit message, virtio_ivshmem.c is for ACRN hypervisor based VM whereas in Kconfig Driver for QEMU IVSHMEM'
Unify them into Acrn ivshmem.

    In drivers/virtio/virtio_guest_shm.c and drivers/virtio/virtio_ivshmem.c
        - In probe function, vi_dev is not freed on pcim_enable_device failure.
        - In probe function, vi_data is not freed neither in err_enable nor in err label
        In virtio_ivshmem_remove function, vi_dev>priv not freed?
    In drivers/ivshm/ivshm_guest_shm.c,
        In probe function, idev is not freed on all failure exits
        In ivshmem_init_msix function, idev->msix_entries is not freed on all failure exits
        Check whether the registered regions are also released on all possible failure exits
For the memory need to free in drivers, i have double checked. It's all allocated by devm_kalloc/devm_kcalloc which do not need explicitly free.

    In drivers/ivshm/ivshm.c,
        cdev is not freed on failures in ivshm_major_init
Add cdev_del function.

    Have we modified the file drivers/virtio/virtio_guest_shm.c? If so whats the change we did in addition to the one done by Jan Kiszka jan.kiszka@siemens.com??
Basically, everything has been changed. The data structure involved in virtio over shared memory has been changed to a layout based on virtio config space itself. The shared memory interface used in the lower layer has been changed from jailhouse's ivshmem v2 to ivshmem and qnx shm, while the overall structure of the code and the memory allocation part have been retained

    param(unsigned int, max_vfs, ~0, 0400) \
Removed the '~'.

    Summary should be <72 characters and should be only 1 line
    Limit the max number of characters in the commit message to 72characters/line
    For workaround patches, please prefix [W/A]
    For patches which needs to be reverted in future after proper fix, commit message needs to be prefixed with [REVERT]
Checked all commit messages.

      Commit summary not proper for 13th patch
      Subject: [PATCH 13/14] =?UTF-8?q?QNX:=20x86=EF=BC=9Aremoving=20unsupported?=
      =?UTF-8?q?=20feature?=
      MIME-Version: 1.0
      Content-Type: text/plain; charset=UTF-8
      Content-Transfer-Encoding: 8bit
Rewrite the Summary.

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

YuanyuanZhao609 commented:
hi Jeevaka, thank you for your comments.

All patches checked with ./scripts/checkpatch.pl, and errors have resolved. 

    'As per commit message, virtio_ivshmem.c is for ACRN hypervisor based VM whereas in Kconfig Driver for QEMU IVSHMEM'
Unify them into Acrn ivshmem.

    In drivers/virtio/virtio_guest_shm.c and drivers/virtio/virtio_ivshmem.c
        - In probe function, vi_dev is not freed on pcim_enable_device failure.
Fixed.
        - In probe function, vi_data is not freed neither in err_enable nor in err label
        In virtio_ivshmem_remove function, vi_dev>priv not freed?
    In drivers/ivshm/ivshm_guest_shm.c,
        In probe function, idev is not freed on all failure exits
        In ivshmem_init_msix function, idev->msix_entries is not freed on all failure exits
        Check whether the registered regions are also released on all possible failure exits
For rest memory need to free in drivers, i have double checked. It's all allocated by devm_kalloc/devm_kcalloc which do not need explicitly free.

    In drivers/ivshm/ivshm.c,
        cdev is not freed on failures in ivshm_major_init
Add cdev_del function.

    Have we modified the file drivers/virtio/virtio_guest_shm.c? If so whats the change we did in addition to the one done by Jan Kiszka jan.kiszka@siemens.com??
Basically, everything has been changed. The data structure involved in virtio over shared memory has been changed to a layout based on virtio config space itself. The shared memory interface used in the lower layer has been changed from jailhouse's ivshmem v2 to ivshmem and qnx shm, while the overall structure of the code and the memory allocation part have been retained

    param(unsigned int, max_vfs, ~0, 0400) \
Removed the '~'.

    Summary should be <72 characters and should be only 1 line
    Limit the max number of characters in the commit message to 72characters/line
    For workaround patches, please prefix [W/A]
    For patches which needs to be reverted in future after proper fix, commit message needs to be prefixed with [REVERT]
Checked all commit messages.

    Commit summary not proper for 13th patch
    Subject: [PATCH 13/14] =?UTF-8?q?QNX:=20x86=EF=BC=9Aremoving=20unsupported?=
    =?UTF-8?q?=20feature?=
    MIME-Version: 1.0
    Content-Type: text/plain; charset=UTF-8
    Content-Transfer-Encoding: 8bit
Rewrited the Summary.

kaushlen commented on drivers/virtio/virtio_guest_shm.c:
```diff
@@ -0,0 +1,198 @@
+				(attach_list & (1u << i)) ? "at" : "de");
+	}
+	vi_data->attach_list = attach_list;
+	return ret;
```
Why to take separate variable ret for this. lets directly do return IRQ_HANDLED;

YuanyuanZhao609 replied:
OK. Changed.

kaushlen commented on drivers/virtio/virtio_guest_shm.c:
```diff
@@ -0,0 +1,198 @@
+	if (ret) {
+		dev_err(&pci_dev->dev, "failed to register device\n");
+		pci_disable_device(pci_dev);
+		put_device(&vi_dev->vdev.dev);
```
should be not free the vi_dev memory here, which you allocated at start?

YuanyuanZhao609 replied:
Go to 'err_enable' to deal it.

kaushlen commented on drivers/virtio/virtio_ivshmem.c:
```diff
@@ -0,0 +1,182 @@
+		dev_err(&pci_dev->dev, "failed to register device\n");
+		pci_disable_device(pci_dev);
+		put_device(&vi_dev->vdev.dev);
+		return ret;
```
should we not free the memory vi_dev ?

YuanyuanZhao609 replied:
Same as above.

kaushlen commented on drivers/ivshm/ivshm_guest_shm.c:
```diff
@@ -0,0 +1,170 @@
+
+	err = pcim_iomap_regions(pdev, BIT(0), "ivshm_guest_shm");
+	if (err)
+		return err;
```
Why we are not undoing the stuff in case of error? like pci_disable_device. You might want to look at other places in this function for similar stuffs.

YuanyuanZhao609 replied:
Go to 'out_disable_device' to handle it coherently.

kaushlen commented on arch/x86/kernel/apic/apic.c:
```diff
@@ -324,6 +324,19 @@ int lapic_get_maxlvt(void)
+{
+	if (hypervisor_is_type(X86_HYPER_QNX))
+		get_option(&str, &qnx_lvtt_scale_factor);
+	return 1;
```
If we are not relying on any of the return value from function, why to return hardcoded 1?
Can we not make it void?

YuanyuanZhao609 replied:
This is a callback for linux __setup macro.

/*
 * NOTE: __setup functions return values:
 * @fn returns 1 (or non-zero) if the option argument is "handled"
 * and returns 0 if the option argument is "not handled".
 */
#define __setup(str, fn)						\
	__setup_param(str, fn, fn, 0)

So i think it should return a value.

kaushlen replied:
Can we put a comment like callback function. are you sure you always want to return 1 from this function?

kaushlen commented on drivers/gpu/drm/i915/i915_debugfs.c:
```diff
@@ -833,7 +833,33 @@ static const struct file_operations i915_forcewake_fops = {
+}
+
+static int
+i915_force_provisioning_vfs_get(void *data, u64 *val)
```
Do we need this function. Everything is hardcoded inside this.

YuanyuanZhao609 replied:
If there is a function, user will get 0 from the file. If set the function to NULL, i think it may return an error.
But refer to i915_drop_caches_get as an example,  return a '0'.
    static int
    i915_drop_caches_get(void *data, u64 *val)
    {
	    *val = DROP_ALL;
    
	    return 0;
    }

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Dependent pr program name is not compatable this pr's program name, for more details please check tracked_on

sysopenci commented:
Android CI has started Engineering Build for this issue ,Please check the linked Tracked-On issue/Android CI Web  for more details.

sysopenci commented:
SUCCESS: Android CI has completed Engineering Build for this issue.Please check the linked Tracked-On issue/Android CI Web  for more details.

sysopenci commented:
Android CI has started MERGE Build for this pr ,Please check the linked Tracked-On issue/Android CI Web  for more details.

sysopenci commented:
Android CI has completed MERGE Build for this pr, build is FAILURE. Please check the linked Tracked-On issue/Android CI Web  for more details.
